openhands/llm/llm.py

```diff
@@ -122,33 +133,58 @@ def __init__(
             top_p=self.config.top_p,
         )
 
+        if self.vision_is_active():
+            logger.debug('LLM: model has vision enabled')
+
         completion_unwrapped = self._completion
 
         def attempt_on_error(retry_state):
+            """Custom attempt function for litellm completion."""
             logger.error(
-                f'{retry_state.outcome.exception()}. Attempt #{retry_state.attempt_number} | You can customize these settings in the configuration.',
+                f'{retry_state.outcome.exception()}. Attempt #{retry_state.attempt_number} | You can customize retry values in the configuration.',
                 exc_info=False,
             )
             return None
 
-        @retry(
-            reraise=True,
-            stop=stop_after_attempt(self.config.num_retries),
-            wait=wait_random_exponential(
+        def custom_completion_wait(retry_state):
+            """Custom wait function for litellm completion."""
+            if not retry_state:
+                return 0
+            exception = retry_state.outcome.exception() if retry_state.outcome else None
+            if exception is None:
+                return 0
+
+            min_wait_time = self.config.retry_min_wait
+            max_wait_time = self.config.retry_max_wait
+
+            # for rate limit errors, wait 1 minute by default, max 4 minutes between retries
+            exception_type = type(exception).__name__
+            logger.error(f'\nexception_type: {exception_type}\n')
+
+            if exception_type == 'RateLimitError':
+                min_wait_time = 60
+                max_wait_time = 240
+            elif exception_type == 'BadRequestError' and exception.response:
```

Comment:
Are you sure we need to hardcode the 429 case like this, aren't multiple decorators supported? If so, it might be worth looking into just writing another decorator for it, with custom values. Though 429 is a bit special indeed.

Reply:
By using this custom method, I don't have to replicate the decorator 3 times (there are the other 2 wrappers, even though unused yet).

Reply:
> I think most of the other errors shouldn't be on the same wait period or not even on retry at all as I'm sceptical, that some of those would/can get better with waiting?

On the first part: yes, I seem to recall using tenacity with two decorators: one with some wait values, another with others. With `retry_if_exception_type`. So we could use a second one for 429, as you noticed that one needs some other values.

On the second part of the quote: sure, the exceptions here should be temporary exceptions, those we are confident enough that they can realistically be gone next time it tries. A fluke or something. I don't know, that's not the case? In think we were using very very few, precisely. If in the meantime we got too many, by all means we can remove any that's inappropriate.